src/utils/elections/states/pennsylvania.py

```diff
@@ -0,0 +1,188 @@
+    headers = {
+        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:137.0) Gecko/20100101 Firefox/137.0",
+        "Accept": "application/json, text/plain, */*",
+        "Referer": "https://www.electionreturns.pa.gov/ReportCenter/Reports",
+    }
+    response = requests.get(url, headers=headers, timeout=10)
+    response.raise_for_status()
+    # the data is returned in a horrible triple json escaped format
+    raw_data = response.json()
+    election_data_json = json.loads(raw_data)[0]["ElectionData"]
+    election_dicts = json.loads(election_data_json)
+    valid_elections = [
+        {
+            "ElectionID": int(election["Electionid"]),
+            "Subtype": election["ElectionType"],
+            "Year": int(election["ElectionYear"]),
+        }
+        for election in election_dicts
+        if (
+            (start_year is None or int(election["ElectionYear"]) >= start_year)
+            and (end_year is None or int(election["ElectionYear"]) <= end_year)
+        )
+    ]
+    return valid_elections
+
+
+def standardize_election_data(election_results: pd.DataFrame) -> pd.DataFrame:
+    """Standardize the election data to a common format.
+
+    Args:
+        election_results: A pandas DataFrame containing the election data.
+            each row is a candidate result for a specific election and
+            county.
+
+    Returns:
+        A pandas DataFrame with standardized column names and types.
+    """
+    # drop missing data
+    election_results = election_results.dropna(
+        subset=["Candidate Name", "Votes", "Office Name", "District Name"]
+    )
+    # Remove commas and convert "Votes" column to integers
+    election_results.loc[:, "Votes"] = (
+        election_results["Votes"].str.replace(",", "").astype(int)
+    )
+    # aggregate election results by candidate
+    aggregated_election_results = (
+        election_results.groupby(
+            [
+                "Office Name",
+                "District Name",
+                "Candidate Name",
+                "Party Name",
+                "election--year",
+                "Election Name",
+            ]
+        )
+        .agg(
+            {
+                "Votes": "sum",
+            }
+        )
+        .reset_index()
+    )
+
+    renaming_dict = {
+        "Office Name": "office_sought",
+        "District Name": "district",
+        "Candidate Name": "candidate--full_name",
+        "Party": "candidate--party",
+        "Votes": "votes_received",
+        "Election Name": "election--name",
+    }
+    aggregated_election_results = aggregated_election_results.rename(
+        columns=renaming_dict
+    )
+    return aggregated_election_results
+
+
+def main() -> None:
+    """Main function to fetch and save election results."""
+    print("Discovering valid elections...")
+    elections = get_all_election_ids(start_year=2020, end_year=2022)
```

Comment:
If you're going to hard-code the start and end years, might be good to add a comment to explain why. Or just add command-line arguments.

Reply:
Should definitely be command-line. Added!